pylint/checkers/base/name_checker/checker.py

```diff
@@ -545,42 +560,60 @@ def _assigns_typevar(node: Optional[nodes.NodeNG]) -> bool:
-        """Check if a TypeVar has a variance and if it's included in the name.
-
-        Returns the args for the message to be displayed.
-        """
+    def _check_typevar(self, name: str, node: nodes.AssignName) -> None:
+        """Check for TypeVar lint violations."""
         if isinstance(node.parent, nodes.Assign):
             keywords = node.assign_type().value.keywords
         elif isinstance(node.parent, nodes.Tuple):
             keywords = (
                 node.assign_type().value.elts[node.parent.elts.index(node)].keywords
             )
 
+        variance = TypeVarVariance.invariant
         for kw in keywords:
             if kw.arg == "covariant" and kw.value.value:
-                if not name.endswith("_co"):
-                    suggest_name = f"{re.sub('_contra$', '', name)}_co"
-                    self.add_message(
-                        "typevar-name-incorrect-variance",
-                        node=node,
-                        args=(name, "covariant", suggest_name),
-                        confidence=interfaces.INFERENCE,
-                    )
-                return
-
-            if kw.arg == "contravariant" and kw.value.value:
-                if not name.endswith("_contra"):
-                    suggest_name = f"{re.sub('_co$', '', name)}_contra"
-                    self.add_message(
-                        "typevar-name-incorrect-variance",
-                        node=node,
-                        args=(name, "contravariant", suggest_name),
-                        confidence=interfaces.INFERENCE,
+                variance = (
+                    TypeVarVariance.covariant
+                    if variance
+                    not in (
+                        TypeVarVariance.contravariant,
+                        TypeVarVariance.double_variant,
                     )
-                return
+                    else TypeVarVariance.double_variant
+                )
+            elif kw.arg == "contravariant" and kw.value.value:
+                variance = (
+                    TypeVarVariance.contravariant
+                    if variance
+                    not in (TypeVarVariance.covariant, TypeVarVariance.double_variant)
+                    else TypeVarVariance.double_variant
+                )
 
-        if name.endswith("_co") or name.endswith("_contra"):
+        if variance == TypeVarVariance.double_variant:
+            self.add_message(
+                "typevar-double-variance",
+                node=node,
+                confidence=interfaces.INFERENCE,
+            )
+        elif variance == TypeVarVariance.covariant and not name.endswith("_co"):
+            suggest_name = f"{re.sub('_contra$', '', name)}_co"
+            self.add_message(
+                "typevar-name-incorrect-variance",
+                node=node,
+                args=(name, "covariant", suggest_name),
+                confidence=interfaces.INFERENCE,
```

Comment:
Is it really inference ? I think this is the kind of check where we don't need it because we check the name directly. (I don't see where we inferred something either). I'm not doing suggestions everywhere in case I'm wrong.

```suggestion
                confidence=interfaces.HIGH,
```

Reply:
No it is. We use inference to determine that we're dealing with a typevar and not with another type of name.